Mathlib/Control/Basic.lean

```diff
@@ -0,0 +1,265 @@
+#align map_bind map_bind
+
+theorem seq_bind_eq (x : m α) {g : β → m γ} {f : α → β} :
+    f <$> x >>= g = x >>= g ∘ f :=
+  show bind (f <$> x) g = bind x (g ∘ f)
+  by rw [← bind_pure_comp, bind_assoc] <;> simp [pure_bind, (· ∘ ·)]
+#align seq_bind_eq seq_bind_eq
+
+#align seq_eq_bind_map seq_eq_bind_mapₓ
+
+@[functor_norm]
+theorem fish_pure {α β} (f : α → m β) : f >=> pure = f := by simp only [(· >=> ·), functor_norm]
+#align fish_pure fish_pure
+
+@[functor_norm]
+theorem fish_pipe {α β} (f : α → m β) : pure >=> f = f := by simp only [(· >=> ·), functor_norm]
+#align fish_pipe fish_pipe
+
+-- note: in Lean 3 `>=>` is left-associative, but in Lean 4 it is right-associative.
+@[functor_norm]
+theorem fish_assoc {α β γ φ} (f : α → m β) (g : β → m γ) (h : γ → m φ) :
+    (f >=> g) >=> h = f >=> g >=> h := by
+  simp only [(· >=> ·), functor_norm]
+#align fish_assoc fish_assoc
+
+variable {β' γ' : Type v}
+
+variable {m' : Type v → Type w} [Monad m']
+
+/-- Takes a value `β` and `List α` and accumulates pairs according to a monadic function `f`.
+Accumulation occurs from the right (i.e., starting from the tail of the list). -/
+def List.mapMAccumR (f : α → β' → m' (β' × γ')) : β' → List α → m' (β' × List γ')
+  | a, [] => pure (a, [])
+  | a, x :: xs => do
+    let (a', ys) ← List.mapMAccumR f a xs
+    let (a'', y) ← f x a'
+    pure (a'', y :: ys)
+#align list.mmap_accumr List.mapMAccumR
+
+/-- Takes a value `β` and `List α` and accumulates pairs according to a monadic function `f`.
+Accumulation occurs from the left (i.e., starting from the head of the list). -/
+def List.mapMAccumL (f : β' → α → m' (β' × γ')) : β' → List α → m' (β' × List γ')
+  | a, [] => pure (a, [])
+  | a, x :: xs => do
+    let (a', y) ← f a x
+    let (a'', ys) ← List.mapMAccumL f a' xs
+    pure (a'', y :: ys)
+#align list.mmap_accuml List.mapMAccumL
+
+end Monad
+
+section
+
+variable {m : Type u → Type u} [Monad m] [LawfulMonad m]
+
+theorem joinM_map_map {α β : Type u} (f : α → β) (a : m (m α)) :
+  joinM (Functor.map f <$> a) = f <$> joinM a := by
+  simp only [joinM, (· ∘ ·), id.def, ← bind_pure_comp, bind_assoc, map_bind, pure_bind]
+#align mjoin_map_map joinM_map_map
+
+theorem joinM_map_joinM {α : Type u} (a : m (m (m α))) : joinM (joinM <$> a) = joinM (joinM a) := by
+  simp only [joinM, (· ∘ ·), id.def, map_bind, ← bind_pure_comp, bind_assoc, pure_bind]
+#align mjoin_map_mjoin joinM_map_joinM
+
+@[simp]
+theorem joinM_map_pure {α : Type u} (a : m α) : joinM (pure <$> a) = a := by
+  simp only [joinM, (· ∘ ·), id.def, map_bind, ← bind_pure_comp, bind_assoc, pure_bind, bind_pure]
+#align mjoin_map_pure joinM_map_pure
+
+@[simp]
+theorem joinM_pure {α : Type u} (a : m α) : joinM (pure a) = a :=
+  LawfulMonad.pure_bind a id
+#align mjoin_pure joinM_pure
+
+end
+
+section Alternative
+
+variable {F : Type → Type v} [Alternative F]
+
+-- [todo] add notation for `Functor.mapConst` and port `functor.map_const_rev`
```

Comment:
Is this a todo for this PR, or for later?

Reply:
Note sure what Anatole had in mind here, but `functor.map_const_rev` is in Lean 3 core: https://leanprover-community.github.io/mathlib_docs/init/control/functor.html#functor.map_const_rev

As for notation, I think it can wait. Either way, I expect this todo is not necessary for this PR.

Reply:
I left this todo because I couldn't directly port the mathlib3 definition of `succeeds` as `(x $> tt) <|> pure ff` due to the absence of `functor.map_const_rev` (which doesn't seem to have an analog in Lean4), and I also noticed that (unlike in mathlib3) we don't have notation for `Functor.mapConst`. 

Indeed it is essentially orthogonal to this PR, and the definition of `succeeds` is defeq to the one in mathlib3, so there's no point in blocking the PR for that imho. But I would expect some of these inconsistencies to come up again (especially the lack of notation for `mapConst`), so maybe it's worth adding these things somewhere (either in core or in mathlib).

Reply:
Okay. Let's keep moving.